### PR TITLE
fix(type): fix parsing array literal and printing struct value

### LINKED ISSUE #4769: Different display format for array/struct with pg
In PG, strings with spaces and empty string are displayed with `""`
```sql
postgres=# select ARRAY['1','','2'];
  array
----------
 {1,"",2}
(1 row)

postgres=# select ARRAY['1',' ','2'];
   array
-----------
 {1," ",2}
(1 row)

postgres=# select ARRAY['1','1 1','2'];
    array
-------------
 {1,"1 1",2}
(1 row)

postgres=# select ARRAY['1',' 1 ','2'];
    array
-------------
 {1," 1 ",2}
(1 row)
```

In risingwave:
```sql
dev=> select ARRAY['1','','2'];
 ?column?
----------
 {1,,2}
(1 row)

dev=> select ARRAY['1',' ','2'];
 ?column?
----------
 {1, ,2}
(1 row)

dev=> select ARRAY['1','1 1','2'];
 ?column?
-----------
 {1,1 1,2}
(1 row)

dev=> select ARRAY['1',' 1 ','2'];
 ?column?
-----------
 {1, 1 ,2}
(1 row)
```

Do we want to follow this behavior?

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

### Fix parsing array literal

fix #4769

- We don't handle double-quotes and null values correctly
```diff
  select '{null, "null"}' :: varchar[];
- {"null","\"null\""}
+ {NULL,"null"}
```

This PR rewrites the parser using recursive descent approach.

### Fix row to text

- Null values should appear empty
- Double quotes are needed if
  - it is an empty string
  - it has leading or tailing whitespaces
  - it contains any of the characters`(),"\`
    - `"` is escaped to `""`
    - `\` is escaped to `\\`

```diff
  select Row(''), Row('"'), Row(' a '), Row('()'), Row(','), Row(NULL), Row(NULL, NULL);
-        ()       (")       ( a )       (())       (,)       (NULL)     (NULL,NULL)
+        ("")     ("""")    (" a ")     ("()")     (",")     ()         (,)
```

### Move parsing logic to common

This PR also moves the parsing logic to a method `ListValue::from_str` in the common crate. The cast function no longer needs to build sub-expression on every call. As a result, the runtime of `cast(varchar) -> int8[]` function decreased from 952.10 µs to 185.21 µs (-80%).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

File: src/tests/regress/data/sql/arrays.sql [hunk 1]
```diff
@@ -440,7 +440,7 @@ select array[];
 -- all of the following should be accepted
 select '{}'::text[];
 --@ select '{{{1,2,3,4},{2,3,4,5}},{{3,4,5,6},{4,5,6,7}}}'::text[];
---@ select '{0 second  ,0 second}'::interval[];
+select '{0 second  ,0 second}'::interval[];
```
Comment (xiangjinwu) [inline src/tests/regress/data/sql/arrays.sql:443]: We do not support different `IntervalStyle`
https://www.postgresql.org/docs/16/runtime-config-client.html#GUC-INTERVALSTYLE

Comment (wangrunji0408) [reply src/tests/regress/data/sql/arrays.sql:443]: But why the style is not the default "postgres" style at this place? 🤔

**Review by xiangjinwu — COMMENTED** [review]
https://www.postgresql.org/docs/16/arrays.html#ARRAYS-IO
https://www.postgresql.org/docs/16/rowtypes.html#ROWTYPES-IO-SYNTAX
#4769

There are many edge cases. It is okay if we cannot resolve them all at once.

The wrong output may have been used in a lot of existing e2e slt tests 😢

File: src/common/src/array/list_array.rs [hunk 2]
```diff
@@ -624,6 +626,85 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+        let trimmed = &trimmed[1..trimmed.len() - 1];
+
+        fn eoi() -> String {
+            "Unexpected end of input.".into()
+        }
+
+        let mut elems = Vec::new();
+        let mut depth = 0;
+        let mut start = 0;
+        let mut chars = trimmed.chars().enumerate();
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/common/src/array/list_array.rs:649]: Oh yes. This is generated by GPT4. Obviously I didn't remind it about unicode. 😅

File: src/common/src/array/list_array.rs [hunk 3]
```diff
@@ -624,6 +626,197 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+                let mut has_escape = false;
+                let s = loop {
+                    match chars.next().ok_or_else(Self::eoi)? {
+                        (_, '\\') => {
+                            has_escape = true;
+                            chars.next().ok_or_else(Self::eoi)?;
+                        }
+                        (i, c @ ',' | c @ '}') => {
+                            let s = self.input[..i].trim_end();
+                            self.input = &self.input[i..];
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:703]: Would like a comment here that `self.input` is to inform parent `parse_array` about the continuation location.

File: src/common/src/array/list_array.rs [hunk 4]
```diff
@@ -624,6 +626,197 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+            /// Parse a datum.
+            fn parse(&mut self) -> Result<Datum, String> {
+                self.skip_whitespace();
+                if self.try_parse_null() {
+                    return Ok(None);
+                }
+                if self.data_type.is_array() {
+                    Ok(Some(self.parse_array()?.into()))
+                } else {
+                    self.parse_value()
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:647]: Given `null` is already handled above, `parse_value` (and `parse_quoted`) should be returning `ScalarImpl` rather than `Datum`.

File: src/common/src/array/list_array.rs [hunk 5]
```diff
@@ -624,6 +626,197 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+                let mut chars = self.input.char_indices();
+                let mut has_escape = false;
+                let s = loop {
+                    match chars.next().ok_or_else(Self::eoi)? {
+                        (_, '\\') => {
+                            has_escape = true;
+                            chars.next().ok_or_else(Self::eoi)?;
+                        }
+                        (i, c @ ',' | c @ '}') => {
+                            let s = self.input[..i].trim_end();
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:702]: panics: `select '{\ }'::varchar[];` due to `trim_end`

File: src/common/src/array/struct_array.rs [hunk 6]
```diff
@@ -434,6 +441,65 @@ impl ToText for StructRef<'_> {
...
+    writer.write_char('"')
+}
+
+/// Remove double quotes from a string.
+/// This is the reverse of [`quote_if_need`].
+#[allow(dead_code)]
+fn unquote_if_need(mut input: &str) -> Cow<'_, str> {
+    if input.starts_with('"') && input.ends_with('"') {
+        input = &input[1..input.len() - 1];
+    }
```
Comment (xiangjinwu) [inline src/common/src/array/struct_array.rs:476]: The rules on inner quotes are different in PostgreSQL with or without outer quotes:
```
test=# select '(1,"a""b")'::z;
     z      
------------
 (1,"a""b")
(1 row)

test=# select '(1,a""b)'::z;
   z    
--------
 (1,ab)
(1 row)
```
It even allows entering and exiting quotes multiple times (unlike in array):
```
test=# select '(1,"a"b"c"d"e")'::z;
     z     
-----------
 (1,abcde)
(1 row)

test=# select '(1,a"b"c"d"e)'::z;
     z     
-----------
 (1,abcde)
(1 row)
```

The precise semantic of this function would depend on how it is invoked by the parser, which is not implemented yet. We have several options here:
* Remove this function until we fully implement varchar -> struct
* Define it to only reverse `quote_if_need` and used by unit tests. Acknowledge in comment it is different from `unescape` needed by `record_in` (aka varchar -> struct)

**Review by xiangjinwu — COMMENTED** [review]
Very close ... Thank you for the effort!

File: src/common/src/array/struct_array.rs (hunk 6 above)
Comment (wangrunji0408) [reply src/common/src/array/struct_array.rs:476]: It sounds too complicated. 🥵 I'd like to remove this function for now.

File: src/common/src/array/list_array.rs [hunk 7]
```diff
@@ -624,6 +626,208 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+            }
+
+            /// Consume the next 4 characters if it matches "null".
+            fn try_parse_null(&mut self) -> bool {
+                if let Some(s) = self.input.get(..4) && s.eq_ignore_ascii_case("null") {
+                    let next_char = self.input[4..].chars().next();
+                    match next_char {
+                        None | Some(',' | '{' | '}' | '\t' | '\n' | '\r' | '\x0B' | '\x0C' | ' ') => {}
+                        // following normal characters
+                        _ => return false,
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:773]:
```
test=# select '{null a}'::varchar[];
  varchar   
------------
 {"null a"}
```

although our own ToText would always quote it.

File: src/common/src/array/list_array.rs [hunk 8]
```diff
@@ -624,6 +626,208 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+                            self.input = &self.input[i..];
+                            if s.trim_end().is_empty() {
+                                return Err(format!("Unexpected \"{c}\" character."));
+                            }
+                            break if has_escape {
+                                // trim whitespaces after unescaping
+                                // consider: "\  " -> " "
+                                let mut s = Self::unescape(s);
+                                s.truncate(s.trim_end().len());
+                                Cow::Owned(s)
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:713]: Unit test failed.

File: src/common/src/array/list_array.rs [hunk 9]
```diff
@@ -624,6 +626,230 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+                            let trimmed = s.trim_end();
+                            // consume the value and leave the ',' or '}' for parent
+                            self.input = &self.input[i..];
+
+                            if trimmed.is_empty() {
+                                return Err(format!("Unexpected \"{c}\" character."));
+                            }
+                            if trimmed.eq_ignore_ascii_case("null") {
+                                return Ok(None);
+                            }
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:712]: Can we move all trim-related logic (incl `let trimmed =`) here in the `!has_escape` (aka `else`) branch?

File: src/common/src/array/list_array.rs [hunk 10]
```diff
@@ -624,6 +626,230 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
...
+                        unescaped.push(c);
+                    }
+                }
+                let l = unescaped[len_after_last_escaped_char..].trim_end().len();
+                unescaped.truncate(len_after_last_escaped_char + l);
+                unescaped
+            }
+
+            /// Consume the next 4 characters if it matches "null".
+            fn try_parse_null(&mut self) -> bool {
```
Comment (xiangjinwu) [inline src/common/src/array/list_array.rs:789]: Given this is only called by `self.data_type.is_array()` now, it can be simplified (or be more specific to prevent misuse). Some ideas include:
* It is an error if not started with `{` or `null`.
* After `null` and `skip_whitespace`, the next char must be `,` or `}`.

^ Not saying we should implement these checks here, but that we can leverage these properties here.

**Review by xiangjinwu — APPROVED** [review]
Did not construct any further problematic cases, but left two suggestions that hopefully improves the readability.

Btw there may be new tests added in main that uses the old output format, and we would need to fix them after merge-main.

File: src/common/src/array/list_array.rs (hunk 10 above)
Comment (wangrunji0408) [reply src/common/src/array/list_array.rs:789]: Even though it is only used by the array branch, I don't think this function should do anything other than what its name suggests, e.g. checking for leading or following chars. These should be left to the upper-level function. 🤔

File: src/common/src/array/list_array.rs (hunk 9 above)
Comment (xiangjinwu) [reply src/common/src/array/list_array.rs]: I meant to clarify the logic, rather than the code structure: trimming is only necessary when there is no escape.

File: src/common/src/array/list_array.rs (hunk 10 above)
Comment (xiangjinwu) [reply src/common/src/array/list_array.rs:789]: As I said, I am not saying we should implement these checks here, but that we can leverage these properties here.

Let me try to rephrase issues I want to address here:
* When peeking the next char, `{` shall not be allowed.
* When peeking the next char, it is better to reuse `skip_whitespace` rather than listing those chars again.
* We have fixed a bug by not calling this function when parsing non-array values. However, its doc / name / implementation still looks like very generic and usable there. Although we can prevent this misuse by having unit test, it is better to make it clear from the code logic - that this function only handles limited case correctly.

File: src/common/src/array/list_array.rs (hunk 9 above)
Comment (wangrunji0408) [reply src/common/src/array/list_array.rs]: Sorry I misunderstood "the branch".

File: src/common/src/array/list_array.rs (hunk 10 above)
Comment (xiangjinwu) [reply src/common/src/array/list_array.rs:789]: > it is better to make it clear from the code logic

To clarify, I do not have an idea of how to achieve this yet. So the suggestions here sound vague. The easiest is just to leave some comments, clearly stating this will cause bug if used when parsing non-array value, and point to the corresponding unit test case as an example.

Comment (wangrunji0408) [reply src/common/src/array/list_array.rs]: Sorry again for my misunderstanding. My intention is to make this function generic and reusable. But it is only useful in parsing arrays. I've added a note to make it clear.

**Comment by wangrunji0408** [comment]
@xiangjinwu Thanks for your patient review. 🥹
